### PR TITLE
Optimization: Reduction Pushdown V2 Part 1

### PR DESCRIPTION
**Pull request by wangandi**
Part of #573.

### Tips for reviewer

Check out the unit tests to see how this is intended to work.

There are also now extensive comments in the code.

The motivation section covers the history of efforts to add reduction pushdown to the optimizer and gives the big picture for what the PR is trying to do. Note that the PR only implements a subset of the big picture. See the description section for what is actually implemented.

#### Aggregations currently supported in PR

* the no aggregation case aggregations
* aggregations on `distinct(<single input expression>)` like `select sum(distinct foo.a)...`

##### Idea

Example query: 
```
select distinct(x.d)
from (
   select * 
   from w, x, y, z 
   where w.a = x.a 
        and x.b = y.b
        and y.c = z.c 
) group by x.b
```

A join can be thought of as a multigraph where vertices are inputs and edges are join constraints. 

The join in the example can be thought of as a graph like so:
```
  w.a = x.a      x.b = y.b    y.c = z.c
w ---------- x ------------ y ----------z
```
For each grouping key k, remove the edges representing join constraints of the form `k = <expression>`.

```
  w.a = x.a                  y.c = z.c
w ---------- x             y ----------z
```
You can push a reduction down to each connected component and then join the reduces along the removed constraints. Thus, the query after reduction pushdown is

```
select reduce1.d
from (
   (select x.b, distinct(x.d) as d from w, x where w.a = x.a group by x.b) reduce1
   (select y.b from w, x where y.c = z.c group by y.b) reduce2
   where reduce1.b = reduce2.b 
)
```

### Motivation

It may be helpful to consult the [arrangement developer docs](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/arrangements.md) where I break down the costs of each kind of reduce arrangements.

#### Reduction Pushdown V1

Was PR #5396. 
A much longer write-up of this can be found [here](https://docs.google.com/document/d/1v-ebzEWGOfflKXIveUzKEP8iBX3y5zh57O3Hb40ry6E/edit#heading=h.6f6sx9r8cc1v).

##### Original idea, explained via example query: 

```
SELECT sum(ol_quantity) AS a
FROM item, orderline
WHERE ol_i_id = i_id
GROUP BY ol_w_id;
```

For each aggregation, see which input the expression comes from. Try to push a reduction down on that particular input.  

```
SELECT sum(partial_sum) AS a
FROM item, 
    (SELECT ol_w_id, ol_i_id, sum(ol_quantity) partial_sum
    FROM orderline GROUP BY ol_w_id, ol_i_id) inner_reduce
WHERE ol_i_id = i_id
GROUP BY ol_w_id;
```

If the aggregation is independent of any input, such as count(*), try to push it down to an input with other aggregations, or failing that, the first input. If the aggregation comes from more than one input, try to break it into multiple aggregations. See (2) under the following heading: https://docs.google.com/document/d/1v-ebzEWGOfflKXIveUzKEP8iBX3y5zh57O3Hb40ry6E/edit#heading=h.13v7brkrbhju

##### Issues

* There have not been any known situations where a user could have benefited a partial reduction pushdown.
* This approach does not support pushing down the reduce when there are no aggregations, e.g.
  ```
  select distinct foo_a
  from foo, bar
  where foo.foo_a = bar.bar_a
  ```
* The cost of a Reduce is usually around `2*(the number of groups)`, so reduction pushdown usually did not change the memory cost of the outer reduce.  Thus, the new inner reduces were often just additional overhead memory-wise. It was difficult to tell whether or not an inner reduce sufficiently reduces the number of rows to be worthwhile. 
* For aggregations like count(*), it was unclear whether pushing it down on the first input was the best way to go.
* The code could introduce unnecessary maps. 

  For example, the output type of a sum is one size larger than the input type of a sum.

  Back to the original example query. If `ol_quantity` is an Int32, then `sum(ol_quantity)` is supposed to be an Int64. 
  `sum(partial_sum)` will be type Decimal, so Reduction Pushdown V1 will add a downcast to preserve the expected type:  
  ```
  SELECT CAST(sum(partial_sum) AS int64) AS a
  FROM item, 
      (SELECT ol_w_id, ol_i_id, sum(ol_quantity) partial_sum
      FROM orderline GROUP BY ol_w_id, ol_i_id) inner_reduce
  WHERE ol_i_id = i_id
  GROUP BY ol_w_id;
  ```
  If it turns out that (ol_w_id, ol_i_id) is a unique key, Reduce Elision will happen:
  ```
  SELECT CAST(sum(partial_sum) AS int64) AS a
  FROM item, 
      (SELECT ol_w_id, ol_i_id, CAST(ol_quantity AS int64) partial_sum
      FROM orderline) inner_reduce
  WHERE ol_i_id = i_id
  GROUP BY ol_w_id;
  ```
  and now there are two unnecessary maps.

##### Idea

The original inspiration comes via @asenac in #7707
```
select foo_a
from foo, bar
where foo.foo_a = bar.bar_a
group by foo_a
```

The set of columns from `foo` used in the join (`{foo_a}`) is a subset of the grouping columns (`{foo_a}`). Thus, the outer reduce can be pushed down to both `foo` and `bar` like so:

```
select foo_a
from (select foo_a from foo group by foo_a) reduced_foo,
(select bar_a from bar group by bar_a) reduced_bar
where reduced_foo.foo_a = reduced_bar.bar_a
```

If the outer reduce contains any aggregation on `distinct(<single-input expressions>`, when the reduce gets pushed down, the aggregation goes to the input the aggregation references.

```
select
  sum(distinct foo_b) as distinct_foo_b_sum,
  count(distinct bar_b) as distinct_bar_b_count
from foo, bar
where foo.foo_a = bar.bar_a
group by foo_a
```

can be rewritten as:

```
select distinct_foo_b_sum, distinct_bar_b_count
from 
  (select foo_a, sum(distinct foo_b) as distinct_foo_b_sum from foo group by foo_a) reduced_foo,
  (select bar_a, count(distinct bar_b) as distinct_bar_b_count from bar group by bar_a) reduced_bar
where reduced_foo.foo_a = reduced_bar.bar_a
```

If the outer reduce contains certain non-distinct aggregations, partial results can be computed across the two inner reduces:

```
select
  sum(foo_b) as foo_b_sum,
  count (*) as count,
  sum(foo_b * bar_b) as sum_b_product
from foo, bar
where foo.foo_a = bar.bar_a
group by foo_a
```

can be rewritten as:

```
select 
  foo_b_partial_sum * bar_count as foo_b_sum, 
  foo_count * bar_count as count, 
  foo_b_partial_sum * bar_b_partial_sum as sum_b_product
from
  (select foo_a, sum(foo_b) as foo_b_partial_sum, count(*) as foo_count)  reduced_foo,
  (select bar_a, sum(bar_b) as bar_b_partial_sum, count(*) as bar_count)  reduced_bar
where reduced_foo.foo_a = reduced_bar.bar_a
```

##### Advantages of V2 over V1

* Hopefully easier to understand than V1.

* Supports pushing down reduces with no aggregate expressions, i.e. distincts.

* Limits application of reduction pushdown to a subset of cases where it is more likely to help while still supporting supports the cases where we have seen reduction pushdown make life better for a user. If a user is writing a query like:
  ```
  select sum(foo_b)
  from foo, bar
  where foo.foo_a = bar.bar_a
  group by foo_a
  ```
  Then it is more likely that an inner reduce grouping on `foo_a` will reduce the size of the join enough to be worthwhile.

* Should have better performance characteristics. Suppose we have the query 
  ```
  select sum(foo_b)
  from foo, bar
  where foo.foo_a = bar.bar_a
  group by foo_a
  ```
  and both `foo_a` and `bar_a` have the same set of distinct values. 

  Then the V1 approach
  ```
  select sum(foo_b_partial_sum)
  from
     (select foo_a, sum(foo_b) as foo_b_partial_sum from foo group by foo_a) reduced_foo,
     bar
  from reduced_foo, reduced_bar
  where reduced_foo.foo_a = reduced_bar.bar_a
  ```
  will have the same number and size of reduce arrangements as the V2 approach:
  ```
  select foo_b_partial_sum * bar_count
  from
     (select foo_a, sum(foo_b) as foo_b_partial_sum from foo group by foo_a) reduced_foo,
     (select bar_a, count(*) as bar_count group by bar_a) reduced_bar
  where reduced_foo.foo_a = reduced_bar.bar_a
  ```
  But the V2 approach would shrink the join more.

* Is great for reduces around a cross join. The empty set is always a subset of the grouping columns. For example, this query with an inner join + a cross join.
  ```
  select sum(foo_b) as foo_b_sum
  from foo, bar, baz
  where foo.foo_a = bar.bar_a
  group by foo.foo_c
  ```
  Will be transformed to:
  ```
  select foo_b_partial_sum * count as foo_b_sum
  from
    (select sum(foo_b) as foo_b_partial_sum
     from foo, bar 
     where foo.foo_a = bar.bar_a 
     group by foo.foo_c) foo_bar_join,
    (select count(*) as count from baz) collapsed_baz
  ```

### Checklist

- [ ] This PR has adequate test coverage.
- [ ] This PR adds a release note for any user-facing behavior changes.

### CODE REVIEW

**Comment by wangandi** [comment]
@frankmcsherry requested in person that I break down the diffs between V1 and V2 by category.

Diff in the math used:
[In this document](https://docs.google.com/document/d/1xrBJGGDkkiGBKRSNYR2W-nKba96ZOdC2mVbLqMLjJY0), I describe what is happening in V1 and V2 in terms of the 3 conditions required to perform a full reduction pushdown as described in [Section 3.1 in Galindo-Lagoria and Joshi (2001)](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.563.8492&rep=rep1&type=pdf). The tl;dr version is that.
* In V2, we no longer try to work around condition 1 (Columns from R involved in join constraints are a subset of the group by keys).
* Both V1 and V2 try to workaround condition 2 (The key of T is a subset of the group by keys.), but V2 uses a different workaround compared to V1.
There is no change to the workaround for condition 3 (The columns involved in the aggregation all belong to R.).

Features V2 takes into account that were missing from V1:
* Pushing down a GroupBy with no aggregations. 
* A less arbitrary decision for pushing down zero-input aggregations like `count(*)`

Engineering approach diff:
@frankmcsherry brought up the possibility that in the future, we may want to allow the user to toggle the expansiveness of the cases where reduction pushdown happens. V2 does not prevent a future where we allow the user to do reduction pushdown when condition 1 is not fulfilled; a future can be, for example, combining the condition 1 workaround in V1 with the condition 2 workaround in V2.

I see Reduction Pushdown V2 as changing the order of reduction pushdown work to start off with reduction pushdown only happening with the subset of cases that is the most likely to be useful. This has the advantage of:
* Not having to do the additional work of hiding reduction pushdown behind a hint.
* Better testing. I worry that hiding reduction pushdown behind a hint will cause reduction pushdown to be inadequately tested. Currently, changes we make to transforms often lead to unexpected edge case behavior that is discovered by other team members after the fact. I would prefer to have support for only cases that do not need to be hidden on main for a while that way we have some time to collectively get a better sense of how this transform would interact with others.
* Faster landing of code that is likely to have immediate benefit for users.

**Comment by philip-stoev** [comment]
Item No#1. Not sure if this can only happen as part of this transform, but I am getting the following fragment in the Explain:

```
# 2021-12-03T13:30:29  %1 =
# 2021-12-03T13:30:29  | Get materialize.public.t2 (<U>)
# 2021-12-03T13:30:29  | Project ()
# 2021-12-03T13:30:29 -| Distinct group=()
# 2021-12-03T13:30:29 -| ArrangeBy ()
```

So my question is, why do we `ArrangeBy` on a `Distinct group=()` -- my understanding is that `Distinct group=()` is essentially one group with a `NULL` key or something along those lines? Is there a need for it to be Arranged?

**Comment by philip-stoev** [comment]
Item No#2. This query:

```
EXPLAIN    SELECT  FROM t2 a2  WHERE  2  IN (  SELECT f1  FROM t1   )    AND  EXISTS (  SELECT  FROM t1  WHERE  TRUE  OR a2  IS  NULL  )     ;```

produces a 3-table cross-join in this branch:

```
-%3 =
-| Join %0 %1 %2
-| | implementation = Differential %0 %1.() %2.()
```

I guess it is an overall win, but that particular 3-table cross join is not there in the prior commit. Contrast with the same query but where the unnecessary `TRUE OR` predicate has been removed -- in this case, the 3-table join in this branch is not a cross join:

```
 %5 =
-| Join %2 %3 %4 (= #0 #2) (= #1 #3)
-| | implementation = Differential %2 %3.(#0, #1) %4.()
```

**Review by philip-stoev — APPROVED** [review]
I am able to see this optimization in action -- the several dozen query plans that I examined all seem to fit the desired pattern and appear to be beneficial -- In addition to the expected move of Distinct to before the join, I observe joins being folded and/or eliminated altogether.

There were no wrong results or panics.

**Comment by wangandi** [comment]
> So my question is, why do we `ArrangeBy` on a `Distinct group=()` -- my understanding is that `Distinct group=()` is essentially one group with a NULL key or something along those lines? Is there a need for it to be Arranged?

`Distinct group=()` is essentially one group with a key consisting of an empty row. It does not need to be `Arranged`. That said, the presence of the `ArrangeBy` is not harmful. Currently, `ArrangeBy` means "Arrange if not already Arranged". Since `Distinct group=()` is already arranged, the ArrangeBy is a no-op.

**Comment by wangandi** [comment]
> I guess it is an overall win, but that particular 3-table cross join is not there in the prior commit. Contrast with the same query but where the unnecessary `TRUE OR` predicate has been removed.

I'm confused. I would think that a `TRUE AND` predicate would be unnecessary, but not `TRUE OR`. In the Item No 2. query, I would consider the `a2 IS NULL` predicate to be unnecessary instead of `TRUE OR`.

**Comment by philip-stoev** [comment]
> > I guess it is an overall win, but that particular 3-table cross join is not there in the prior commit. Contrast with the same query but where the unnecessary `TRUE OR` predicate has been removed.
> 
> I'm confused. I would think that a `TRUE AND` predicate would be unnecessary, but not `TRUE OR`. In the Item No 2. query, I would consider the `a2 IS NULL` predicate to be unnecessary instead of `TRUE OR`.

I am sorry,  you are right. I spaced out on this so please disregard the entire explanation.

It is just that the `TRUE OR` plan appeared to be more complex than was warranted. Anyway, it seems to be a problem outside of the scope of this PR -- in other contexts I also saw `TRUE AND` and `TRUE OR` not being fully optimized away.

**Review by asenac — APPROVED** [review]

File: src/transform/src/reduction_pushdown.rs [hunk 1]
```diff
@@ -80,7 +120,288 @@ impl ReductionPushdown {
...
+                .flat_map(|expr| old_join_mapper.lookup_inputs(expr)),
+        );
+        // b) Extract the set of components that covers the inputs.
+        let (components_to_connect, other): (Vec<_>, Vec<_>) = components
+            .into_iter()
+            .partition(|(inputs, _)| inputs.iter().any(|i| inputs_to_connect.contains(i)));
+        // c) Union the components. Add the current constraint to
+        let mut connected_component = Vec::new();
+        let mut all_constraints = vec![equivalence];
+        for (mut componentn, mut constraintsn) in components_to_connect {
```
[the remark this answers is not in the document]
Comment (wangandi) [reply src/transform/src/reduction_pushdown.rs]: Renamed to `component` and `constraint_set`. Originally, I was using std::iter::fold, so I had `component1` as a starting value and folded in `componentn`.

**Review by aalexandrov — APPROVED** [review]
LGTM modulo some minor remarks from my side.

After reading through the code, my general impression is that we need to do a lot of bookkeeping in order to re-map column references correctly after the rewrite.

I was therefore wondering (a) if this is a general problem and (b) if using global column naming will be helpful to simplify this areas of the code.

File: src/transform/src/lib.rs [hunk 2]
```diff
@@ -328,7 +328,6 @@ impl Optimizer {
                     Box::new(crate::map_lifting::LiteralLifting::default()),
                 ],
             }),
-            Box::new(crate::reduction_pushdown::ReductionPushdown),
```
[the remark this answers is not in the document]
Comment (wangandi) [reply src/transform/src/lib.rs:331]: If you scroll up in the same file, you'll see ReductionPushdown runs on line 292, as part of the logical transforms. The call in the physical transforms here is being permanently removed. It turns out this line is not doing anything on main at the moment anyway.

File: src/transform/tests/testdata/reduction-pushdown [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+%0 =
+| Get x (u1)
+
+%1 =
+| Get y (u2)
+
+%2 =
+| Join %0 %1 (= #0 #2)
+| | implementation = Unimplemented
+| Distinct group=(#3, #3)
```
[the remark this answers is not in the document]
Comment (wangandi) [reply src/transform/tests/testdata/reduction-pushdown:291]: The code to simplify the `Distinct group=(#3, #3)` already exists in the transform `ProjectionExtraction`. I have added a version of the test to show that if you run the full set of optimizations, the redundancy will be removed.

I'm assuming that the followup issue would be whatever we decide for solving the "Inefficient transformation" problem.

File: src/transform/tests/testdata/reduction-pushdown [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+(reduce (join [(get x) (get y) (get z)] [[#3 #5]]) [#5] [])
+----
+----
+%0 =
+| Get x (u1)
+| Distinct group=()
+
+%1 =
+| Get y (u2)
+| Distinct group=(#1)
```
[the remark this answers is not in the document]
Comment (wangandi) [reply src/transform/tests/testdata/reduction-pushdown:350]: The way the code currently works is:
For each group_key, see if there is an equivalent key in the component. If yes, add it to the group keys of the reduce on that component. 

In the previous test, there are two group keys, both of which is equivalent to something belonging to the component (`#3` is equivalent to `#3` and `#5` is equivalent to `#3`), hence the duplication. Here, there is only one group key.

File: src/transform/src/reduction_pushdown.rs [hunk 5]
```diff
@@ -80,7 +120,302 @@ impl ReductionPushdown {
...
+
+    // 3b) Deduce the aggregates that each reduce needs calculate in order to
+    // reconstruct each aggregate in the old reduce.
+    for agg in aggregates {
+        if let Some(component) =
+            lookup_corresponding_component(&agg.expr, &old_join_mapper, &input_component_map)
+        {
+            if agg.distinct == false {
+                // TODO: support non-distinct aggs
+                return None;
```
[the remark this answers is not in the document]
Comment (wangandi) [reply src/transform/src/reduction_pushdown.rs:303]: Added a follow-up issue describing support for an expanded set of aggregates at #9604. Link is also being added to the code comment.

File: src/transform/src/reduction_pushdown.rs [hunk 6]
```diff
@@ -7,12 +7,45 @@
...
+//! Suppose there are two inputs R and T being joined. According to
+//! [Galindo-Legaria and Joshi (2001)](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.563.8492&rep=rep1&type=pdf),
+//! a full reduction pushdown to R can be done if and only if:
+//! 1) Columns from R involved in join constraints are a subset of the group by keys.
+//! 2) The key of T is a subset of the group by keys.
+//! 3) The columns involved in the aggregation all belong to R.
+//!
+//! In our current implementation:
+//! * We abide by condition 1 to the letter.
+//! * We work around condition 2 by pushing the reduction down to both inputs.
```
[the remark this answers is not in the document]
Comment (wangandi) [reply src/transform/src/reduction_pushdown.rs:30]: I'm not sure what is being asked for here, but I've tried to clarify this comment.

**Comment by wangandi** [comment]
> After reading through the code, my general impression is that we need to do a lot of bookkeeping in order to re-map column references correctly after the rewrite.
> 
> I was therefore wondering (a) if this is a general problem and (b) if using global column naming will be helpful to simplify this areas of the code.

I'm not sure what you are specifically envisioning when you say "global column naming"?

I do feel that "whether we can reduce bookkeeping" should be considered as part of the "Maintainability" theme. I created the `JoinInputMapper` in hopes of reducing bookkeeping when doing transforms on joins, where the problem is most severe. Personally, I feel that bookkeeping is also a bit of a problem when it comes to `Map`s because frequently you have to do `scalars[column - input.arity()]`.

**Pull request merged**